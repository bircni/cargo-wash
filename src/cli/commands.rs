use crate::{
    cli::Commands,
    utils::{self},
};

impl Commands {
    /// Show the output for the command selected
    pub fn show(&self) -> anyhow::Result<()> {
        match self {
            Self::Stats(opts) => {
                let (projects, _) = opts.check_args()?;
                utils::show_stats(&projects);
            }
            Self::Clean(opts) => {
                let (projects, dry_run) = opts.check_args()?;
                utils::run_clean(&projects, dry_run, opts.exclude.as_ref())?;
            }
        }

        Ok(())
    }
}
