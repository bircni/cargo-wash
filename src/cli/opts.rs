use clap::Parser;
use parking_lot::RwLock;
use rayon::iter::{ParallelBridge as _, ParallelIterator as _};
use std::fs;
use std::path::PathBuf;
use std::sync::Arc;

use crate::{data::Project, logic};

/// Represents the command line options
#[derive(Parser, Default, Clone)]
pub struct Opts {
    /// Path to the directory from which to start the search for Rust projects
    /// default is `.`
    #[clap(long, short)]
    pub path: Option<PathBuf>,
    /// Exclude the provided folder from the size calculation and cleaning
    /// default is `None` - you can specify multiple folders separated by `,`
    #[clap(long, short)]
    pub exclude: Option<String>,
}

impl Opts {
    pub fn check_args(&self) -> anyhow::Result<Vec<Project>> {
        let projects: Arc<RwLock<Vec<Project>>> = Arc::new(RwLock::new(vec![]));
        let path = logic::sanitize_path_input(self.path.clone())?;

        if path.is_dir() {
            logic::check_project(&path, self.exclude.as_ref()).map(|p_opt| {
                if let Some(project) = p_opt {
                    projects.write().push(project);
                }
            })?;

            match fs::read_dir(&path) {
                Ok(entries) => {
                    entries.flatten().par_bridge().for_each(|entry| {
                        let pathbuf = entry.path();
                        if pathbuf.is_dir() {
                            match logic::check_project(&pathbuf, self.exclude.as_ref()) {
                                Ok(p_opt) => {
                                    if let Some(project) = p_opt {
                                        projects.write().push(project);
                                    }
                                }
                                Err(error) => log::warn!("Error checking project: {error}"),
                            }
                        }
                    });
                }
                Err(error) => log::warn!("Error reading directory: {error}"),
            }
        } else {
            anyhow::bail!("The provided path is not a directory.");
        }

        Ok(projects.read().to_vec())
    }
}
