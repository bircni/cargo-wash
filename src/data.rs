use core::fmt::{self, Display};
use std::path::{Path, PathBuf};

/// Represents the size unit
/// B = Bytes
/// KB = Kilobytes
/// MB = Megabytes
/// GB = Gigabytes
#[derive(strum_macros::Display, Clone, Copy, PartialEq, PartialOrd, Debug)]
pub enum SizeUnit {
    B,
    GB,
    KB,
    MB,
}

/// Represents a size in bytes with a unit
#[derive(Clone, Copy, PartialEq, PartialOrd, Debug)]
pub struct Size {
    pub unit: SizeUnit,
    pub value: f64,
}

impl Size {
    pub const fn new(value: f64, unit: SizeUnit) -> Self {
        Self { unit, value }
    }

    /// Returns the size in bytes
    #[expect(clippy::cast_possible_truncation, reason = "Ok here")]
    #[expect(clippy::cast_sign_loss, reason = "Ok here")]
    pub fn size_in_bytes(&self) -> u64 {
        let multiplier = match self.unit {
            SizeUnit::B => 1,
            SizeUnit::KB => 1024,
            SizeUnit::MB => 1024 * 1024,
            SizeUnit::GB => 1024 * 1024 * 1024,
        };
        (self.value * f64::from(multiplier)) as u64
    }

    /// Converts a size in bytes to a `Size` struct
    #[expect(clippy::cast_precision_loss, reason = "Ok here")]
    pub fn to_size(bytes: u64) -> Self {
        const KB: u64 = 1024;
        const MB: u64 = KB * 1024;
        const GB: u64 = MB * 1024;

        if bytes >= GB {
            Self::new(bytes as f64 / GB as f64, SizeUnit::GB)
        } else if bytes >= MB {
            Self::new(bytes as f64 / MB as f64, SizeUnit::MB)
        } else if bytes >= KB {
            Self::new(bytes as f64 / KB as f64, SizeUnit::KB)
        } else {
            Self::new(bytes as f64, SizeUnit::B)
        }
    }
}

impl Display for Size {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:.2} {}", self.value, self.unit)
    }
}

/// Represents a Rust project
/// with its name, path, and size
#[derive(Clone, Debug)]
pub struct Project {
    pub name: String,
    pub path: PathBuf,
    pub size: Size,
}

impl Project {
    pub fn new<P: AsRef<Path>>(name: &str, path: P, size: u64) -> Self {
        Self {
            name: name.to_owned(),
            path: path.as_ref().to_path_buf(),
            size: Size::to_size(size),
        }
    }
}
