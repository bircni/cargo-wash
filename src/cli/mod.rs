pub mod opts;

use clap::Parser;
use opts::Options;

use crate::commands::{clean, execute, stats};

/// Represents the available commands
#[derive(Parser)]
#[command(author, version, about)]
pub enum Commands {
    /// Execute `cargo clean` on all projects to remove build artifacts
    Clean(Options),
    /// Print statistics about all Rust projects in the directory
    Stats(Options),
    /// Execute different commands on all rust projects in the directory
    Execute(Options),
}

impl Commands {
    /// Run the command based on the provided options
    pub fn run(&self) -> anyhow::Result<()> {
        match self {
            Self::Stats(opts) => {
                let projects = opts.check_args()?;
                stats::show(&projects);
            }
            Self::Clean(opts) => {
                let projects = opts.check_args()?;
                clean::run(&projects, opts.exclude.as_ref())?;
            }
            Self::Execute(opts) => {
                let projects = opts.check_args()?;
                execute::run(&projects, opts)?;
            }
        }

        Ok(())
    }
}
