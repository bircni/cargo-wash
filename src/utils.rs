use std::{fs, process::Command};

use comfy_table::Table;
use log::debug;
use rayon::iter::{IntoParallelRefIterator, ParallelIterator};

use crate::data::{Language, Project, Size};

pub fn total_size_of_projects(projects: &[Project]) -> u64 {
    projects
        .par_iter()
        .map(|project| project.size.size_in_bytes())
        .sum()
}

/// Recursively calculate the size of a folder
pub fn get_folder_size<P: AsRef<std::path::Path>>(dir: P) -> anyhow::Result<u64> {
    if !dir.as_ref().exists() {
        return Ok(0);
    }
    let mut total_size = 0;

    for entry in fs::read_dir(dir)? {
        let entry = entry?;
        let path = entry.path();
        let metadata = fs::metadata(&path)?;

        if metadata.is_dir() {
            total_size += get_folder_size(&path)?;
        } else {
            total_size += metadata.len();
        }
    }

    Ok(total_size)
}

pub fn get_project<P: AsRef<std::path::Path>>(dir: P) -> Language {
    let path = dir.as_ref();

    if path.join("Cargo.toml").is_file() {
        return Language::Rust;
    } else if path.join("package.json").is_file() {
        return Language::NodeJS;
    }

    Language::Other
}

pub fn show_stats(projects: &[Project]) {
    let mut sorted_projects: Vec<Project> = projects.to_vec();
    sorted_projects.sort_by(|a, b| a.name.to_lowercase().cmp(&b.name.to_lowercase()));
    let mut table = Table::new();
    table.set_header(vec!["Project", "Size", "Path", "Language"]);

    for project in sorted_projects {
        table.add_row(vec![
            &project.name,
            &project.size.to_string(),
            &project.path.to_string_lossy().to_string(),
            &project.language.to_string(),
        ]);
    }

    table.add_row(vec![
        "Total",
        &Size::to_size(total_size_of_projects(projects)).to_string(),
    ]);

    println!("{table}");
}

pub fn print_status(projects: &[Project], cleaned: &[Project], dry_run: bool) {
    let str = if dry_run {
        "Would have cleaned"
    } else {
        "Cleaned"
    };
    let used_projects = if dry_run { projects } else { cleaned };
    let total_size = total_size_of_projects(used_projects);
    println!(
        "{str} {} ({} Projects)\nProjects: {}",
        Size::to_size(total_size),
        projects.len(),
        used_projects
            .iter()
            .map(|p| p.name.clone())
            .collect::<Vec<String>>()
            .join(", ")
    );
}

pub fn run_clean(projects: &[Project], dry_run: bool) -> anyhow::Result<()> {
    let mut cleaned_projects = vec![];

    for project in projects {
        // Check if the project is Rust or Node.js based on its language type
        if dry_run {
            debug!("Would clean: {:?}", project.name);
            continue;
        }
        match project.language {
            Language::Rust => {
                debug!("Running `cargo clean` for project: {:?}", project.name);

                let result = Command::new("cargo")
                    .arg("clean")
                    .current_dir(&project.path)
                    .output();

                match result {
                    Ok(output) => {
                        if output.status.success() {
                            cleaned_projects.push(project.clone());
                        } else {
                            anyhow::bail!(
                                "Failed to clean {}: {}",
                                project.name,
                                String::from_utf8_lossy(&output.stderr)
                            );
                        }
                    }
                    Err(e) => anyhow::bail!("Failed to clean {}: {}", project.name, e),
                }
            }
            Language::NodeJS => {
                debug!("Removing node_modules for project: {:?}", project.name);

                if std::fs::remove_dir_all(project.path.join("node_modules")).is_ok() {
                    cleaned_projects.push(project.clone());
                } else {
                    anyhow::bail!("Failed to remove node_modules for {}", project.name);
                }
            }
            Language::Other => {
                debug!("Skipping project: {:?}", project.name);
            }
        }
    }

    print_status(projects, &cleaned_projects, dry_run);

    Ok(())
}
