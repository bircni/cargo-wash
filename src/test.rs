use std::path::PathBuf;

use anyhow::Context;
use clap::ColorChoice;
use clap::Command;

use crate::cli;
use crate::data;
use crate::utils;

fn snapshot_test_cli_command(app: Command, cmd_name: &str) -> anyhow::Result<()> {
    let mut app = app
        .color(ColorChoice::Never)
        .version("0.0.0")
        .long_version("0.0.0");

    let mut buffer = Vec::new();
    app.write_long_help(&mut buffer)?;
    let help_text = std::str::from_utf8(&buffer)?;

    if insta::_macro_support::assert_snapshot(
        cmd_name.into(),
        help_text,
        env!("CARGO_MANIFEST_DIR"),
        "cli-cmd",
        module_path!(),
        file!(),
        line!(),
        "help_text",
    )
    .is_err()
    {
        anyhow::bail!("Snapshot test failed for command: {}", cmd_name);
    }

    for app in app.get_subcommands() {
        if app.get_name() == "help" {
            continue;
        }

        snapshot_test_cli_command(app.clone(), &format!("{cmd_name}-{}", app.get_name()))?;
    }
    Ok(())
}

#[allow(clippy::expect_used)]
#[test]
fn cli_snapshot() {
    use clap::CommandFactory;

    insta::with_settings!({
        snapshot_path => "../tests/snapshots",
    }, {
        snapshot_test_cli_command(
            super::Commands::command().name("cargo_wash"),
            "cargo_wash",
        ).expect("Failed to run snapshot test");
    });
}

#[test]
fn test_clean_path() {
    assert!(utils::clean_path(Some(PathBuf::from("/"))).is_ok());
    assert!(utils::clean_path(Some(PathBuf::from("."))).is_ok());
    assert!(utils::clean_path(Some(PathBuf::from(".."))).is_ok());
    assert!(utils::clean_path(Some(PathBuf::from("test"))).is_ok());
}

#[test]
fn test_get_folder_size() -> anyhow::Result<()> {
    let size = utils::get_folder_size("src")?;
    assert!(size > 0);
    Ok(())
}

#[test]
fn test_get_project() {
    assert_eq!(utils::get_language("."), cli::Language::Rust);
    assert_eq!(utils::get_language("src"), cli::Language::Other);
}

#[test]
fn test_check_project() -> anyhow::Result<()> {
    let opts = cli::Opts::default();
    let res = opts
        .check_project(&PathBuf::from("../cargo-wash"), Some(cli::Language::Rust))?
        .context("Project was `None`")?;
    assert!(res.size > data::Size::to_size(0));
    assert!(res.language == cli::Language::Rust);
    assert!(res.name == "cargo-wash");
    assert!(res.path == PathBuf::from("../cargo-wash"));
    Ok(())
}

#[cfg(not(target_os = "windows"))] //Windows does not allow deleting the current executable
#[test]
fn test_run_clean() -> anyhow::Result<()> {
    let opts = cli::Opts::default();
    let (projects, _) = opts.check_args()?;
    utils::run_clean(&projects, false)?;
    Ok(())
}
