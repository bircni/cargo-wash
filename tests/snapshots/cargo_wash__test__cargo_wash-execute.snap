---
source: src/test.rs
expression: help_text
---
Execute different commands on all rust projects in the directory

Usage: execute [OPTIONS]

Options:
  -p, --path <PATH>
          Path to the directory from which to start the search for Rust projects default is `.`

  -e, --exclude <EXCLUDE>
          Exclude the provided folder from the size calculation and cleaning default is `None` - you can specify multiple folders separated by `,`

  -c, --command <COMMAND>
          The command to execute on the Rust projects It is only used for the `execute` command

          Possible values:
          - build:  Execute `cargo build`
          - check:  Execute `cargo check`
          - doc:    Execute `cargo doc`
          - clean:  Execute `cargo clean`
          - run:    Execute `cargo run`
          - test:   Execute `cargo test`
          - bench:  Execute `cargo bench`
          - update: Execute `cargo update`
          - none:   No command, used for default or no operation

  -h, --help
          Print help (see a summary with '-h')

  -V, --version
          Print version
